File: source/formview.php
<?php
session_start();
?>
<!DOCTYPE html>
<html>
<head>
<style>.error{color: red;}</style>
<title>Generic table CRUD screen</title>
</head>
<body>
<?php

	echo "Hello1";

	// read the application parameters from a text file into an associative array
	$parameters = array();
	foreach( file("parameters.txt") as $line ){
		list($key, $value) = explode("=", str_replace(array("\r", "\n"), "", $line));
		$parameters[$key] = $value;
	}

	// default the language to "en" if default language not supported
	$language = substr($_SERVER["HTTP_ACCEPT_LANGUAGE"], 0, 2);
	$supports = ["en", "es", "fr"];	// supported languages
	$language = in_array($language, $supports) ? $language : "en";

	// record id and table name come from either of GET or POST
	if ( empty($id) ){
		$id = empty($_GET["id"]) ? $_POST["id"] : $_GET["id"];
	}

	// tablename should be a consistent value so make it a session variable
	if ( empty($_SESSION["tablename"]) ){
		$_SESSION["tablename"] = empty($_GET["tablename"]) ? $_POST["tablename"] : $_GET["tablename"];
		// check it's populated because it's required
		if ( empty($_SESSION["tablename"]) ){
			die("Need to know which table to perform CRUD action on!");
		}
	}


	// open a connection to mySQL using $parameters
	$conn = new mysqli($parameters["servername"], $parameters["username"], $parameters["password"], $parameters["database"]);
	if ( $conn->connect_errno ){
		die("Connection failed : " . $conn->connect_errno);
	}

	if ( ! empty($_POST) ){

		// sanitise anything being POSTed in to thwart script injection attacks
		foreach( $_POST as $name => $value ){
			$_POST[$name] = htmlspecialchars($value);
		}

		var_dump($_POST);

		// it's either create or update
		if ( empty($_POST["id"]) ){
			foreach( $_POST as $name => $value ){
				$fields = sprintf("%s,%s", $fields, $name);
				$values = sprintf("%s,%s", $values, $value);
			}
			$sql = sprintf("INSERT INTO %s (%s) VALUES(%s);", $_SESSION["tablename"], $fields, $values);
		}else{
			foreach( $_POST as $name => $value ){
				$pairs  = sprintf("%s=%s,", $name, $value);
			}
			$sql = sprintf("UPDATE %s SET %s WHERE id = '%s';", $_SESSION["tablename"], $pairs, $_POST["id"]);
		}
/*
		if ( $conn->query($sql) === TRUE ){
		}else{
			die("Error in insert/update : " . $sql);
		}
*/
		//echo $sql;

		

	}

	// fetch the table metadata
	$sql = sprintf("
		SELECT information_schema.columns.column_name, data_type, referenced_table_schema, referenced_table_name, referenced_column_name, character_maximum_length, is_nullable, column_type 
		FROM information_schema.columns 
		LEFT OUTER JOIN information_schema.key_column_usage ON 
			information_schema.key_column_usage.table_schema = information_schema.columns.table_schema AND 
			information_schema.key_column_usage.table_name = information_schema.columns.table_name AND 
			information_schema.key_column_usage.column_name = information_schema.columns.column_name 
		WHERE 	information_schema.columns.table_schema = '%s' AND 
			information_schema.columns.table_name = '%s' AND 
			information_schema.columns.column_name <> 'id';", 
		$parameters["database"], $_SESSION["tablename"]);
	$metadata = $conn->query($sql);
	if ( $metadata->num_rows == 0 ){
		die("Unable to retrieve metadata for table " . $_SESSION["tablename"]);
	}

	// retrieve any language specific input field labels for the columns in database.table into an associative array of column_name=>label
	$sql = sprintf(
		"SELECT column_name, html_input_label FROM column_labels WHERE table_schema='%s' AND table_name='%s' AND language_code = '%s';",
		$parameters["database"], $_SESSION["tablename"], $language
	);
	$results = $conn->query($sql);
	while ( $row = $results->fetch_assoc() ){
		$labels[$row["column_name"]] = $row["html_input_label"];
	}

	printf("<h2>Create, Retrieve, Update, and Delete - %s</h2>", $_SESSION["tablename"]);

	echo "<p><span class='error'>* signifies a required field</span></p>";

	// retrieve the record identified by $id ( unless we're creating a new record in which case $id has no value )
	$sql = sprintf("SELECT * FROM %s WHERE id = %s;", $_SESSION["tablename"], $id);
	$data = $conn->query($sql);
	if ( $data->num_rows > 0 ){
		$record = $data->fetch_assoc();	// (field=>value,...)
		$data->free_result();
	}

	// loop back to self on post
	printf("<form action= '%s' method='post'>", $_SERVER['PHP_SELF']);

	// easier to hard code these fields as to word around them in the code
	printf("<input id='id' name='id' type='hidden' value='%s'>", $record["id"]);

	// loop through the metadata (each row is one column of $table)
	while ( $column = $metadata->fetch_assoc() ){

		if ( $column["IS_NULLABLE"] == "NO" ){
			$required = "required";
		}else{
			$required = "";
		}

		// convert each column's SQL datatype into an appropriate HTML input field type

		switch ( $column["DATA_TYPE"] ){
			case "datetime":
				$type = "date";
				break;
			case "int":
				$type = "number";
				break;
			case "varchar":
				$type = "text";
				break;
			default:
				$type = "text";
		}

		// labels either come from the column_labels table or default to the column name itself

		if ( ! empty($labels[$column["COLUMN_NAME"]]) ){
			printf("<label for='%s'>%s</label><br>", $column["COLUMN_NAME"], $labels[$column["COLUMN_NAME"]]);			
		}else{
			printf("<label for='%s'>%s</label><br>", $column["COLUMN_NAME"], $column["COLUMN_NAME"]);
		}

		// regardless of their data type (and to ensure referential integrity) foreign keys appear as a drop down list of appropriate values

		if ( ! empty($column["referenced_table_schema"]) ){

			// foreign key lookups are defined in the foreign_key_lookups table

			$sql = sprintf("
				SELECT * 
				FROM foreign_key_lookups 
				WHERE primary_table_schema = '%s' AND primary_table_name = '%s' AND primary_column_name = '%s';", 
				$parameters["database"], $_SESSION["tablename"], $column["COLUMN_NAME"]
			);

			$results = $conn->query($sql);

			if ( $results->num_rows > 0 ){

				$foreign_key_lookup = $results->fetch_assoc();

				$sql = sprintf(
					"SELECT %s.%s.%s AS option_value, %s.%s.%s AS option_inner FROM %s.%s;", 
					$foreign_key_lookup["foreign_table_schema"], $foreign_key_lookup["foreign_table_name"], $foreign_key_lookup["html_option_value_column"],
					$foreign_key_lookup["foreign_table_schema"], $foreign_key_lookup["foreign_table_name"], $foreign_key_lookup["html_option_inner_column"],
					$foreign_key_lookup["foreign_table_schema"], $foreign_key_lookup["foreign_table_name"]
				);

				$options = $conn->query($sql);

			}else{

				// if there are no lookups for schema.table.column default to using the foreign key itself as both value and inner

				$sql = sprintf(
					"SELECT %s AS option_value, '' AS option_inner FROM %s.%s", 
					$column["referenced_column_name"], $column["referenced_table_schema"], $column["referenced_table_name"]
				);

				$options = $conn->query($sql);

			}

			printf("<select id='%s' name='%s' value='%s'>", $column["COLUMN_NAME"], $column["COLUMN_NAME"], $record[$column["COLUMN_NAME"]]);

			while ( $option = $options->fetch_assoc() ){
				if ( $option["option_value"] == $record[$column["referenced_column_name"]] ){
					printf("<option value='%s' selected>%s</option>", $option["option_value"], $option["option_inner"]);
				}else{
					printf("<option value='%s'>%s</option>", $option["option_value"], $option["option_inner"]);
				}
			}

			echo "</select>";	

			$options->free_result();

		}elseif ( $column["DATA_TYPE"] == "enum" ){	// enumerated types appear as radio buttons
			// convert information_schema.column_type to an array
			$enums = str_getcsv(str_replace(["enum(",")"], ["",""], $column["COLUMN_TYPE"]), ",", "'");
			foreach ( $enums as $enum ){
				if ( $record[$column["COLUMN_NAME"]] === $enum ){
					$checked = "checked";
				}else{
					$checked = "";
				};
				printf("<input type='radio' id='%s' name='%s' %s value='%s' %s>", $enum, $column["COLUMN_NAME"], $required, $enum, $checked);
				printf("<label for='%s'>%s</label>", $enum, $enum);
			}
		}elseif( $column["DATA_TYPE"] == "datetime" ){
				$value = date_format(date_create($record[$column["COLUMN_NAME"]]), "Y-m-d");
				printf("<input type='%s' id='%s' name='%s' %s value='%s'>", $type, $column["COLUMN_NAME"], $column["COLUMN_NAME"], $required, $value);
		}else{
				$value = $record[$column["COLUMN_NAME"]];
				printf("<input type='%s' id='%s' name='%s' %s value='%s'>", $type, $column["COLUMN_NAME"], $column["COLUMN_NAME"], $required, $value);
		}

		if ( ! empty($required) ){
			echo "<span class='error'> * </span>";
		}

		echo "<br>";

	}

	echo "<input type='submit' value='submit'>";

	echo "</form>";

	$conn->close();

?>
</body>
</html>

File: source/formview.pl
#!C:\Strawberry\perl\bin\perl.exe

use CGI;
use CGI::Session;
use Config::Properties;
use Data::Dumper;
use DBD::Oracle qw(:ora_types);
use DBI;
use HTML::Template;
use strict;
use Switch;
use warnings;

$CGI::APPEND_QUERY_STRING = 1;		# append GET/POST parameters so Vars returns both
$CGI::POST_MAX = 1024;			# avoid DoS attacks by limiting the size of a POST

my $cgi = CGI->new;

# should have logged in so should have a cookie from which we get the session id
my $sid = $cgi->cookie('CGISESSID') || $cgi;

# create a session using the sid
my $session = CGI::Session->load("driver:File", $sid, {Directory=>'/tmp'}) or die CGI::Session->errstr();

# session parameters are stored in a file named for the session id, stored  in the server's /tmp directory 
my $username = $session->param('username');

# if any of these conditions apply we're going to redirect the client to the login page 
if ( ! defined $sid || $session->is_expired || ! defined $username ){
	print $cgi->redirect('login.pl');
	exit;	
}

# always do this regardless of its being a GET or a POST
print "Content-Type: text/html\n\n";

# get GET/POST parameters as a hash of name/value pairs
my $params = $cgi->Vars;

# convert all the %params keys to UC for consistency between the script, the template, and the stored procedure(s)

map { 
	if ( $_ =~ qr/[a-z]+/mp ){ 				# if the key contains any lowercase letters then...
		$params->{uc $_} = $params->{$_}; 	# 	create a new uppercase key with the same value as the lower
		delete($params->{$_}); 				# 	delete the lowercase key so there's no ambuguity in the hash
	}  										# end if
} keys %{$params};

my $table = uc $params->{'TABLE'};	# this is case sensitive in stored procedures	
delete($params->{'TABLE'});			# delete the table key/value pair from the hash

my $action = uc $params->{'ACTION'};	# this is case sensitive in stored procedures	
delete($params->{'ACTION'});			# delete the action key/value pair from the hash

my @columns = keys(%{$params});	# all field/column names sans TABLE and ACTION

# load the database properties from properties/database.properties

my $filename = sprintf("%s/properties/database.properties", $ENV{DOCUMENT_ROOT});

open my $pfh, '<', $filename or die "Unable to open $filename file!";

my $properties = Config::Properties->new();

$properties->load($pfh);

my $dsn = $properties->getProperty("database.dsn");

my $schema = uc $properties->getProperty("database.username");

# if we're not logged in as the DBA then we're using a proxy off of the DBA

if ( $username ne $schema ){
	$username = sprintf("%s[%s]", $schema, $username);
}

my $password = $properties->getProperty("database.password");

# get the database connection (one per session, auto commit disabled)

my $dbh = DBI->connect($dsn, $username, $password, { RaiseError => 1, AutoCommit => 0 }) || die(DBI->errstr());

# set the date format for the templates/procedures

$dbh->do("ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD'");

# retrieve primary and foreign key metadata for given schema/table

my $sql = "BEGIN $schema.retrieve_metadata('$schema', '$table', :CURSOR); END;";

my $sth = $dbh->prepare($sql);

$sth->bind_param_inout(":CURSOR", \my $ref, 0, {ora_type => ORA_RSET});

$sth->execute();

my $keys_meta_data = $ref->fetchall_hashref('COLUMN_NAME');

my @message;

if ( $ENV{'REQUEST_METHOD'} eq 'POST' ) {

	# estabish some common variables to draw upon while composing the appropriate SQL statement for $action

	#my @keys = keys(%{$keys_meta_data});	# key column names
	my @holders = ();			# values for placeholders

	# regardless of it's being an INSERT, UPDATE, or DELETE always use ? placeholdes for values
	# DBI will automatically escape these thus preventing the possibility of SQL injection attack

	switch ( $action ){

		case 'INSERT' {

			# INSERT INTO table (field, field,...) VALUES (?, ?,...)

			my $insertables;	# hash of params of non primary key columns

			while ( my ($key, $value) = each %{$params} ){
				if ( $keys_meta_data->{$key}->{'CONSTRAINT_TYPE'} ne 'P' ){
					$insertables->{$key} = $params->{$key};
				}
			}

			$sql = "INSERT INTO $schema.$table (" . join(', ', keys %{$insertables}) . ') VALUES (' . '?,' x ((keys %{$insertables}) - 1) . '?)';

			@holders = values(%{$insertables});

		}		

		case 'UPDATE' {

			# UPDATE table SET field = ? WHERE field = ? AND field = ?

			my $setters = '';	# the SET field = ? clause
			my $filters = '';	# the WHERE field = ? clause

			#my %values = %{$params};
			my @values = ();

			# compose the SQL UPDATE statement

			while ( my ($index, $column) = each @columns ){
				if ( $keys_meta_data->{$column}->{'CONSTRAINT_TYPE'} eq 'P' ){								# if the column is a primary key then
					$filters = "$filters $column = ? AND";													# 	add a placeholder to the WHERE clause
					push(@holders, $params->{$column});														# 	add the value of the key to the array of key values
					#delete(%values{$column});																#	delete the name/value pair from the params hash
				}else{																						# else
					$setters = $index != @columns - 1 ? "$setters $column = ?, " : "$setters $column = ? ";	# 	all other field(s) go in the SET clause
					push(@values, $params->{$column});														#	add the value to the list of placeholder values
				} 																							# end if
			}

			$sql = "UPDATE $schema.$table SET $setters WHERE $filters 1 = 1";					# add 1 = 1 to round out the trailing AND in the WHERE clause

			@holders = (@values, @holders);

		}

		case 'DELETE' {

			# DELETE FROM table WHERE field = ? AND field = ?

			$sql = "DELETE FROM $schema.$table WHERE ";

			while ( my ($index, $column) = each @columns ){
				$sql = $index != @columns - 1 ? "$sql $column = ? AND" : "$sql $column = ?";
			}

			@holders = values(%{$params});

		}

	}

	print "<p>$sql<p>@holders";
	
	# execute the SQL query, passing the appropriate values for the placeholders

	@message = ({MESSAGE => [{ level => 'green', title => 'Success', value => 'Your changes have been committed to the database.' }]});

	$sth = $dbh->prepare($sql);

	$sth->execute(@holders);

	$dbh->commit();

	if ( $action eq 'INSERT' ){
		$action = 'SELECT';
		$params->{$columns[0]} = $dbh->last_insert_id(undef, $schema, $table, $columns[0]);
		print "***** INSERT - $columns[0] = $params->{$columns[0]} *****";
	}

}

#if ( $ENV{'REQUEST_METHOD'} eq 'GET' ) {

	# compose a SQL SELECT statement from the QUERY_STRING name/value pairs (the first being the table name)
	# remembering to always use ? placeholdes as DBI will automatically escape these to prevent SQL injection

	my $record;

	if ( $action eq 'INSERT' ){

		$sql = "SELECT * FROM $schema.$table FETCH NEXT 1 ROWS ONLY";

		$sth = $dbh->prepare($sql);

		$sth->execute();

		$record = $sth->fetchrow_hashref;	# ref to hash of result record 

		# we need a blank record for an INSERT

		foreach my $key ( keys %{$record} ){
			$record->{$key} = '';
		}

	}else{

		$sql = "SELECT * FROM $schema.$table WHERE ";

		while ( my($index, $column) = each @columns ) {
			$sql = $index != keys(%{$params}) - 1 ? "$sql $column = ? AND " : "$sql $column = ?";
		}

		$sth = $dbh->prepare($sql);

		# get the record 

		print "<p>$sql";

		print "<p>", Dumper(values%{$params});

		$sth->execute(values%{$params});

		$record = $sth->fetchrow_hashref;	# ref to hash of result record 

	}

	# get the column metadata

	my $column_names = $sth->{NAME};	# ref to array of column names

	my $column_types = $sth->{TYPE};	# ref to array of column types

	my $html_type;				# the HTML input type of the column SQL data type

	my @fields;				# an array of hashes (one per input) to be passed to the template

	while ( my ($column_index, $column_name) = each @{$column_names} ) {

		# determine the HTML input type from the column's SQL data type

		switch ( @{$column_types}[$column_index] ) {
			case  3	{ $html_type = 'number'; }
			case  8	{ $html_type = 'number'; }
			case 12	{ $html_type = 'text'; }
			case 93	{ $html_type = 'date'; }
			else	{ $html_type = 'text'; }
		}

		# populate the array of field/input metadata hashes to be passed to the template

		switch( $keys_meta_data->{$column_name}->{'CONSTRAINT_TYPE'} ){

			# primary keys are immutable and so their inputs are readonly

			case 'P' {
				push(@fields, { $column_name => [{ column => $column_name, disabled => 'readonly', type => $html_type, value => $record->{$column_name} }]});
			}

			# foreign keys' inputs are a drop down of valid values

			case 'R' {

				# the value/label pairs for the given table and column are read from stored procedure DROPDOWN
				
				# any errors with uninitialized values, check the store procedure 'dropdown' has a handler for the table/colum

				$sql = "BEGIN $schema.dropdown('$table', '$column_name', :CURSOR); END;";

				$sth = $dbh->prepare($sql) or die $sth->errstr;

				$sth->bind_param_inout(":CURSOR", \my $rsh, 0, {ora_type => ORA_RSET}) or die $sth->errstr;

				$sth->execute() or die $sth->errstr;

				my $ref = $rsh->fetchall_arrayref();

				my @options = ();

				while ( my ($value, $array) = each @{$ref} ){
					my ($value, $label) = @{$array};
					if ( $record->{$column_name} eq $value ){
						push(@options, {'label'=>$label, selected => 'selected', 'value'=>$value});
					}else{
						push(@options, {'label'=>$label, 'value'=>$value});
					}
				}

				push(@fields, { $column_name => [{ column => $column_name, options => \@options }]});			

			}

			# all the rest are inputs of whatever data type the COLUMN is

			else {
				push(@fields, { $column_name => [{ column => $column_name, type => $html_type, value => $record->{$column_name} }]});
			}

		}

	}

	# load/open the HTML template in htdocs/templates

	$filename = sprintf("%s/templates/formview.tmpl", $ENV{DOCUMENT_ROOT});

	my $template = HTML::Template->new(filename => $filename, die_on_bad_params => 0);	# die_on_bad_params => 0 so we ignore 'Attempt to set nonexistent parameter' errors

	$template->param(FIELDS => \@fields);		# pass the array of hashes by reference
	$template->param(TABLE => $table);
	$template->param(HEADING => \@message);		# "<div class='w3-panel w3-red'><h3>Warning!</h3><p>$message</p></div>"
	$template->param(ACTION => $action);
 
	# send the template output to the client device (stdout)
	print $template->output;

	$sth->finish;

	$dbh->disconnect;

#}

exit;